pasture-tools/src/bin/render/renderer.rs

```diff
@@ -0,0 +1,689 @@
+            usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+            format: swapchain_format,
+            width: size.width,
+            height: size.height,
+            present_mode: wgpu::PresentMode::Fifo,
+        };
+
+        surface.configure(&device, &surface_config);
+        let depth_view = Self::create_depth_texture(&surface_config, &device);
+
+        Renderer {
+            adapter,
+            device,
+            queue,
+            surface,
+            uniform_bind_group_layout,
+            uniform_bind_group,
+            ubo,
+            surface_config,
+            pipeline: render_pipeline,
+            pipeline_layout,
+            cam: Camera::new(),
+            cam_controller: Box::new(FPSCameraController::new()),
+            model_mat: Matrix4::identity(),
+            depth_view,
+            // cam_controller: Box::new(ArcballCameraController::new()),
+            last_frame: Instant::now(),
+            gpu_point_buffer: pgpu::GpuPointBufferPerAttribute::new(),
+            point_count: 0,
+        }
+    }
+
+    fn create_depth_texture(
+        config: &wgpu::SurfaceConfiguration,
+        device: &wgpu::Device,
+    ) -> wgpu::TextureView {
+        let depth_texture = device.create_texture(&wgpu::TextureDescriptor {
+            size: wgpu::Extent3d {
+                width: config.width,
+                height: config.height,
+                depth_or_array_layers: 1,
+            },
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: Self::DEPTH_FORMAT,
+            usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+            label: None,
+        });
+
+        depth_texture.create_view(&wgpu::TextureViewDescriptor::default())
+    }
+
+    pub fn load_points(&mut self, reader: &mut LASReader) {
+        let default_layout = reader.get_default_point_layout();
+        println!("point layout: {}", default_layout);
+
+        let point_count = reader.remaining_points();
+        let layout = PointLayout::from_attributes(&[
+            POINT_ATTRIB_3D_F32,
+            attributes::COLOR_RGB,
+        ]);
+
+        self.point_count = point_count as u32;
+        let mut point_buffer = InterleavedVecPointStorage::with_capacity(point_count, layout);
+
+        println!("Reading {} points (this can take a while for large point clouds)", point_count);
+        reader.read_into(&mut point_buffer, point_count).unwrap();
+        println!("Done!");
+
+        let inf = f32::INFINITY;
+        let mut aabb_min = Vector3::new(inf, inf, inf);
+        let mut aabb_max = Vector3::new(-inf, -inf, -inf);
+
+        println!("computing aabb...");
+
+        for position in point_buffer.iter_attribute::<Vector3<f32>>(&POINT_ATTRIB_3D_F32) {
+            let mut pos = position;
+            let swizzle_yz = true;
+            if swizzle_yz {
+                pos = Vector3::new(pos.x, pos.z, pos.y);
```

Comment:
Always switching Y and Z is probably not a good default. Many coordinate systems for point clouds have Z=UP, so this might be confusing. Is there a technical limitation with the camera controls that requires this? I would expect that a more versatile renderer will simply provide means to set a model matrix for a whole point cloud and transform the coordinates at runtime

Reply:
We do the switch here because the camera system (well, graphics APIs) expect Y=UP while all point clouds I've seen have Z=UP. I'm not sure an external model matrix would be a good idea since at the moment we take care to translate/scale the point cloud to make it easy to view and navigate with the camera, should this take the external matrix into account or not? Maybe a bool parameter to load_points whether to switch y and z is enough?